src/Serializer/Upcast/Upcast.php

```diff
@@ -0,0 +1,19 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Patchlevel\EventSourcing\Serializer\Upcast;
+
+/**
+ * @psalm-immutable
+ */
+final class Upcast
+{
+    /**
+     * @param class-string         $class
+     * @param array<string, mixed> $payload
+     */
+    public function __construct(public string $class, public array $payload)
```

Comment:
tried to make it as simple as possible, if the move to 8.1 we could make then even readonly